### PR TITLE
KAFKA-16861: Don't convert to group to classic if the size is larger than group max size.

### PR DESCRIPTION
**Pull request by frankvicky**
Fix the bug where the group downgrade to a classic one when a member leaves, even though the consumer group size is still larger than `classicGroupMaxSize`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by chia7712** [comment]
I noticed that we don't have IT for downgrading to classic group. Maybe we can add IT in this PR?

**Comment by frankvicky** [comment]
Sounds good, I will write an integration test for downgrading to classic group

**Comment by dajac** [comment]
@dongnuo123 is working on adding integration tests for all this work stream. We should wait.

@chia7712 We do already have unit tests but we missed one for this particular case.

**Comment by dajac** [comment]
@dongnuo123 Could you please review this PR?

**Comment by dajac** [comment]
@frankvicky Out of curiosity, how did you find the bug? Regarding the integration tests, if you want to help, we can sync with @dongnuo123.

**Comment by chia7712** [comment]
> Out of curiosity, how did you find the bug? 

@dajac the ticket is created by me (see https://issues.apache.org/jira/browse/KAFKA-16861), and I noticed that issue when reading the new coordinator.

**Comment by chia7712** [comment]
> if you want to help, we can sync with @dongnuo123.

@dongnuo123 Please feel free to make this fix be a part of your PR if you are adding integration tests for all this work stream.

**Comment by dajac** [comment]
@chia7712 We can merge this one with the unit test. That’s fine. I just wanted to ensure that we don’t do duplicate work on the integration tests part of it.

**Comment by frankvicky** [comment]
Hi @dajac,

As @chia7712 mentioned, I am cooperating with @chia7712 on this issue. It seems that @dongnuo123 is working on the integration test for this. If that is the case, please feel free to incorporate this fix into your PR.

Thank you! 😺

**Comment by dajac** [comment]
@frankvicky Thanks for this! As I said, I am fine with merging the fix. We should definitely keep it. I just wanted to signal that there is an ongoing effort from @dongnuo123 to add integration and system tests for the upgrade/downgrade paths. We don’t have do to it here.

**Comment by chia7712** [comment]
@frankvicky @dajac thanks for all your responses. I will take a look at this PR later. @dajac it would be great if this PR can have reviews from you :)

**Comment by dajac** [comment]
Sure. I can review it tomorrow.

**Review by dajac — APPROVED** [review]
LGTM, thanks.

**Comment by dajac** [comment]
@chia7712 Do you want to review it too?

**Comment by chia7712** [comment]
> Do you want to review it too?

I will take a look later!

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java
Comment (chia7712) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java]: It seems this variable is unused.

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for this patch. one small comment is left.

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java
Comment (frankvicky) [reply group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java]: Oops, my bad, I will remove it. 😨

**Review by chia7712 — APPROVED** [review]
LGTM. will merge it if QA pass

**Comment by dajac** [comment]
Merged to trunk and to 3.8.

**Pull request merged**
